### PR TITLE
PXE Provisioning fixtures and test using new infra

### PR DESCRIPTION
**Pull request by ogajduse**
This PR introduces a couple of fixtures that use the brand new workflows on our AAP. The functionality that this PR brings in will allow SatQE to perform PXE provisioning on our infrastructure.
The test contained in this PR aims to use all the fixtures and demonstrate their usage and show Provisioning* component owners how they can be used in future tests or how to use them in the existing tests that will be moved to our old new infra.

New broker release that includes https://github.com/SatelliteQE/broker/pull/159/ needs to be done in order the tests to pass.

### CODE REVIEW

**Review by peterdragun — COMMENTED** [review]
Great work :partying_face:, I can see that there are still parts that need to be implemented, but a lot of work has already been done.

File: tests/foreman/cli/test_provisioning.py [hunk 1]
```diff
@@ -16,7 +16,220 @@
...
+
+from robottelo import constants
+from robottelo.api.utils import enable_rhrepo_and_fetchid
+from robottelo.api.utils import wait_for_tasks
+from robottelo.config import settings
+from robottelo.hosts import ContentHost
+
+
+@pytest.fixture(scope='module')
+def module_location(module_manifest_org, module_target_sat):
```
[the remark this answers is not in the document]
Comment (ogajduse) [reply tests/foreman/cli/test_provisioning.py]: Thank you for the suggestion! I used `module_org_with_manifest` it even allowed me to remove API calls that read LCE and CV.

File: tests/foreman/cli/test_provisioning.py [hunk 2]
```diff
@@ -16,7 +16,220 @@
...
+        kickstart_repository=provisioning_rhel_content.repo,
+        lifecycle_environment=lce,
+        root_pass=host_root_pass,
+        operatingsystem=provisioning_rhel_content.os,
+        ptable=default_partitiontable,
+        subnet=subnet,
+        pxe_loader=pxe_loader,
+    ).create()
+
+    return Box(sat=sat, hostgroup=hostgroup, subnet=subnet, cv=cv, lce=lce)
```
[the remark this answers is not in the document]
Comment (ogajduse) [reply tests/foreman/cli/test_provisioning.py]: All addressed.

File: tests/foreman/cli/test_provisioning.py [hunk 3]
```diff
@@ -37,3 +250,56 @@ def test_rhel_pxe_provisioning_on_libvirt():
         2. Host is registered to Satellite and subscription status is 'Success'
         3. Host can install package from Satellite
     """
+
+
+# @pytest.mark.on_premises_provisioning
+@pytest.mark.parametrize(
+    "provisioning_rhel_content",
+    [v for v in settings.supportability.content_hosts.rhel.versions if re.match('^[7-9]$', str(v))],
```
[the remark this answers is not in the document]
Comment (ogajduse) [reply tests/foreman/cli/test_provisioning.py:53]: Good point! This needs to be addressed. I will do more test runs tomorrow.

File: pytest_fixtures/component/provision_pxe.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        rh_repo = module_target_sat.api.Repository(id=rh_repo_id).read()
+        task = rh_repo.sync(synchronous=False)
+        tasks.append(task)
+        rh_repos.append(rh_repo)
+    for task in tasks:
+        wait_for_tasks(
+            search_query=(f'id = {task["id"]}'),
+            poll_timeout=2500,
+        )
+        task_status = module_target_sat.api.ForemanTask(id=task['id']).poll()
```
Comment (peterdragun) [inline pytest_fixtures/component/provision_pxe.py:68]: wait_for_tasks should return list of task_statuses so no need to call API again.

File: tests/foreman/cli/test_provisioning.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ogajduse) [reply tests/foreman/cli/test_provisioning.py]: Addressed in the latest version.

File: pytest_fixtures/component/provision_pxe.py (hunk 4 above)
Comment (ogajduse) [reply pytest_fixtures/component/provision_pxe.py:68]: `wait_for_tasks` returns a list of tasks. In my case, there is one task in the list. This task still has the result `pending`. That is why I have to do an extra API call. The implementation of `wait_for_tasks` would have to change.

File: tests/foreman/cli/test_provisioning.py [hunk 5]
```diff
@@ -37,3 +43,127 @@ def test_rhel_pxe_provisioning_on_libvirt():
...
+        operatingsystem=module_provisioning_rhel_content.os,
+        subnet=module_provisioning_sat.subnet,
+        host_parameters_attributes=[
+            {'name': 'remote_execution_connect_by_ip', 'value': 'true', 'parameter_type': 'boolean'}
+        ],
+        build=True,  # put the host in build mode
+    ).create(create_missing=False)
+    # Clean up the host to free IP leases on Satellite.
+    # broker should do that as a part of the teardown, putting here just to make sure.
+    request.addfinalizer(host.delete)
```
Comment (mshriver) [inline tests/foreman/cli/test_provisioning.py:98]: :heart:

File: tests/foreman/cli/test_provisioning.py [hunk 6]
```diff
@@ -37,3 +43,127 @@ def test_rhel_pxe_provisioning_on_libvirt():
...
+
+    # TODO: Implement Satellite log capturing logic to verify that
+    # all the events are captured in the logs.
+
+    # Host should do call back to the Satellite reporting
+    # the result of the installation. Wait until Satellite reports that the host is installed.
+    wait_for(
+        lambda: host.read().build_status_label != 'Pending installation',
+        timeout=900,
+        delay=30,
```
Comment (mshriver) [inline tests/foreman/cli/test_provisioning.py:119]: Please reduce the delay here, maybe 10s?

File: tests/foreman/cli/test_provisioning.py [hunk 7]
```diff
@@ -37,3 +43,127 @@ def test_rhel_pxe_provisioning_on_libvirt():
...
+    # addressing hosts using FQDNs, falling back to IP.
+    provisioning_host.hostname = host.ip
+    # Host is not blank anymore
+    provisioning_host.blank = False
+
+    # Wait for the host to be rebooted and SSH daemon to be started.
+    try:
+        wait_for(
+            provisioning_host.connect,
+            fail_condition=lambda res: res is not None,
```
Comment (mshriver) [inline tests/foreman/cli/test_provisioning.py:135]: Might be the first use of lambda in fail_condition that I've seen? Cool!

File: pytest_fixtures/core/broker.py [hunk 8]
```diff
@@ -22,7 +22,13 @@ def _resolve_deploy_args(args_dict):
 def _default_sat(align_to_satellite):
     """Returns a Satellite object for settings.server.hostname"""
     if settings.server.hostname:
-        return Satellite()
+        hosts = Broker(host_classes={'host': Satellite}).from_inventory(
```
Comment (mshriver) [inline pytest_fixtures/core/broker.py:25]: @JacobCallahan thoughts here? Almost seems like we need a classmethod instance constructor for satellite.

**Review by mshriver — CHANGES_REQUESTED** [review]
1-2 small things, this is **great** @ogajduse !

File: pytest_fixtures/core/broker.py (hunk 8 above)
Comment (JacobCallahan) [reply pytest_fixtures/core/broker.py:25]: This is currently the correct pattern

File: tests/foreman/cli/test_provisioning.py [hunk 9]
```diff
@@ -37,3 +43,127 @@ def test_rhel_pxe_provisioning_on_libvirt():
         2. Host is registered to Satellite and subscription status is 'Success'
         3. Host can install package from Satellite
     """
+
+
+# TODO: move the whole test to api/ directory
+
+# @pytest.mark.on_premises_provisioning
```
[the remark this answers is not in the document]
Comment (ogajduse) [reply tests/foreman/cli/test_provisioning.py:50]: Oops, I have to uncomment it.

File: tests/foreman/cli/test_provisioning.py (hunk 7 above)
Comment (ogajduse) [reply tests/foreman/cli/test_provisioning.py:135]: I basically copied this one from @JacobCallahan's `power_control` method. :wink:

File: pytest_fixtures/component/provision_pxe.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+    provisioning_domain_name = f"{gen_string('alpha').lower()}.foo"
+
+    broker_data_out = Broker().execute(
+        workflow="configure-install-sat-provisioning-rhv",
+        artifacts="last",
+        target_vlan_id=settings.provisioning.vlan_id,
+        target_host=sat.name,
+        provisioning_dns_zone=provisioning_domain_name,
+        sat_version=sat.version,
+    )
```
Comment (JacobCallahan) [inline pytest_fixtures/component/provision_pxe.py:126]: :chefskiss

File: pytest_fixtures/component/provision_pxe.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+            },
+        ],
+    ).create()
+
+    return Box(sat=sat, hostgroup=hostgroup, subnet=subnet)
+
+
+@pytest.fixture(scope='module')
+def module_ssh_key_file():
+    (_, layout) = mkstemp(text=True)
```
Comment (JacobCallahan) [inline pytest_fixtures/component/provision_pxe.py:204]: The tuple syntax is implied, so not required here for unpacking
```suggestion
    _, layout = mkstemp(text=True)
```

File: pytest_fixtures/core/xdist.py [hunk 12]
```diff
@@ -24,7 +25,9 @@ def align_to_satellite(worker_id, satellite_factory):
 
     # attempt to add potential satellites from the broker inventory file
     if settings.server.inventory_filter:
-        hosts = Broker().from_inventory(filter=settings.server.inventory_filter)
+        hosts = Broker(host_classes={'host': Satellite}).from_inventory(
+            filter=settings.server.inventory_filter
+        )
```
Comment (JacobCallahan) [inline pytest_fixtures/core/xdist.py:30]: I'm curious as to why this change was made

File: robottelo/hosts.py [hunk 13]
```diff
@@ -191,13 +192,15 @@ def os_version(self):
         return Version(version=version_string)
 
     def setup(self):
-        self.remove_katello_ca()
-        self.execute('subscription-manager clean')
+        if not self.blank:
+            self.remove_katello_ca()
+            self.execute('subscription-manager clean')
```
Comment (JacobCallahan) [inline robottelo/hosts.py:197]: Is this needed because of some state where a katello_ca isn't present causing issues?

File: robottelo/hosts.py [hunk 14]
```diff
@@ -191,13 +192,15 @@ def os_version(self):
...
 
     def teardown(self):
-        if self.nailgun_host:
-            self.nailgun_host.delete()
-        self.unregister()
+        if not self.blank:
+            if self.nailgun_host:
+                self.nailgun_host.delete()
+            self.unregister()
```
Comment (JacobCallahan) [inline robottelo/hosts.py:203]: If you haven't rebased already, please do so as this method has changed and you may not need to modify this method.

File: tests/foreman/cli/test_provisioning.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (JacobCallahan) [reply tests/foreman/cli/test_provisioning.py:47]: Unresolving as this is still commented out

**Review by JacobCallahan — CHANGES_REQUESTED** [review]

File: pytest_fixtures/component/provision_pxe.py (hunk 11 above)
Comment (ogajduse) [reply pytest_fixtures/component/provision_pxe.py]: changed

File: pytest_fixtures/component/provision_pxe.py [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+
+
+@pytest.fixture(scope='module')
+def module_provisioning_capsule(module_target_sat, module_location):
+    """Assings the `module_location` to Satellite's internal capsule and returns it"""
+    capsule = (
+        module_target_sat.api.SmartProxy()
+        .search(query={'search': f'name={module_target_sat.hostname}'})[0]
+        .read()
+    )
```
[the remark this answers is not in the document]
Comment (ogajduse) [reply pytest_fixtures/component/provision_pxe.py]: implemented

File: robottelo/hosts.py (hunk 14 above)
Comment (ogajduse) [reply robottelo/hosts.py:203]: I need to keep the `self.blank` attribute there because I am trying to teardown VM that has FQDN `.` (dot). The dot is there as it is the shortest valid FQDN. It is there because the broker requires AAP to provide the `hostname` of the VM deployed. In my case the VM is completely blank - no system installed on it, and it is stopped - not running at the time it is checked out.
I am open to discussions here with you @JacobCallahan and @mshriver. This is not a thing that we should address in this PR.

Comment (ogajduse) [reply robottelo/hosts.py:207]: I have rebased the branch. Can you please re-review this part?

File: robottelo/hosts.py (hunk 13 above)
Comment (ogajduse) [reply robottelo/hosts.py:201]: see https://github.com/SatelliteQE/robottelo/pull/9246/files#r929309510

File: pytest_fixtures/core/xdist.py (hunk 12 above)
Comment (ogajduse) [reply pytest_fixtures/core/xdist.py:30]: Hmm, good catch. This is a leftover from my attempt to load construct `_default_sat` from inventory. This patch is not needed.
@JacobCallahan Do you prefer to leave it as it was prior this change or keep the patch?

File: tests/foreman/cli/test_provisioning.py (hunk 6 above)
Comment (ogajduse) [reply tests/foreman/cli/test_provisioning.py:108]: done

File: tests/foreman/cli/test_provisioning.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ogajduse) [reply tests/foreman/cli/test_provisioning.py]: Test moved from cli to api directory. Marker uncommented.

**Comment by ogajduse** [comment]
PR rebased, and most of the comments were addressed.

**Review by mshriver — APPROVED** [review]

File: pytest_fixtures/core/xdist.py (hunk 12 above)
Comment (JacobCallahan) [reply pytest_fixtures/core/xdist.py:30]: please leave it as-is. My main concern is that there is no need for a full Satellite object to be constructed here.

File: robottelo/hosts.py (hunk 14 above)
Comment (JacobCallahan) [reply robottelo/hosts.py:207]: @ogajduse i think if you just set `self.subscribed = False` on the instance, you will achieve the same result but easier.

Comment (ogajduse) [reply robottelo/hosts.py:207]: Not quite. `self.subscribed` is a property that `execute`s `subscription-manager register`. Running establishing an SSH session is exactly what I want to avoid.

File: tests/foreman/api/test_provisioning.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+    )
+    host = host.read()
+    assert host.build_status_label == 'Installed'
+
+    # Change the hostname of the host as we know it already.
+    # In the current infra environment we do not support
+    # addressing hosts using FQDNs, falling back to IP.
+    provisioning_host.hostname = host.ip
+    # Host is not blank anymore
+    provisioning_host.blank = False
```
Comment (rplevka) [inline tests/foreman/api/test_provisioning.py:96]: shouldn't this action exist in the tear down section of the `provisioning_host` fixture? I understand it is required only for the teardown (checkin) to succeed and has nothing to do with the test itself

**Review by rplevka — APPROVED** [review]

File: robottelo/hosts.py (hunk 14 above)
Comment (JacobCallahan) [reply robottelo/hosts.py:207]: I was going to recommend a different way, but that's something that can be revisited later. So no worries for this PR
